helix-core/src/comment.rs

```diff
@@ -94,6 +96,212 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+        if commented {
+            changes.push((
+                from + open_pos,
+                from + open_pos + open_token.len() + open_margin as usize,
+                None,
+            ));
+            changes.push((
+                from + close_pos - close_token.len() - close_margin as usize + 1,
+                from + close_pos + 1,
+                None,
+            ));
+        } else {
+            changes.push((from + open_pos, from + open_pos, Some(open.clone())));
+            changes.push((
+                from + close_pos + 1,
+                from + close_pos + 1,
+                Some(close.clone()),
+            ));
+        }
+    }
+    Transaction::change(doc, changes.into_iter())
+}
+
+pub fn split_lines_of_selection(text: RopeSlice, selection: &Selection) -> Selection {
+    let mut ranges = SmallVec::new();
+    for range in selection.ranges() {
+        let (line_start, line_end) = range.line_range(text.slice(..));
+        let mut pos = text.line_to_char(line_start);
+        for line in text.slice(pos..text.line_to_char(line_end + 1)).lines() {
+            let start = pos;
+            pos += line.len_chars();
+            ranges.push(Range::new(start, pos));
+        }
+    }
+    Selection::new(ranges, 0)
+}
+
+pub fn toggle_block_comments_as_line_fallback(
+    text: &Rope,
+    selection: &Selection,
+    tokens: Option<(&str, &str)>,
+) -> Transaction {
+    toggle_block_comments(
+        text,
+        &split_lines_of_selection(text.slice(..), selection),
+        tokens,
+    )
+}
+
+pub enum CommentType {
+    Line,
+    Block,
+    BlockAsLineFallback,
+}
+
+/// Return what CommentType selection should be toggled with
+pub fn comment_type(
+    token: Option<&str>,
+    tokens: Option<(&str, &str)>,
+    text: RopeSlice,
+    selection: &Selection,
+) -> CommentType {
+    let mut lines: Vec<usize> = Vec::with_capacity(selection.len());
+    let mut min_next_line = 0;
+    for selection in selection {
+        let (start, end) = selection.line_range(text);
+        let start = start.max(min_next_line).min(text.len_lines());
+        let end = (end + 1).min(text.len_lines());
+
+        lines.extend(start..end);
+        min_next_line = end;
+    }
+    let split_lines = split_lines_of_selection(text, selection);
+    let (line_commented, block_commented) = match (token, tokens) {
+        (Some(_), Some(tokens)) => (
+            find_block_comments(tokens.0, tokens.1, text, &split_lines).0,
+            find_block_comments(tokens.0, tokens.1, text, selection).0,
+        ),
+        (None, None) => (
+            find_block_comments("/*", "*/", text, &split_lines).0,
+            find_block_comments("/*", "*/", text, selection).0,
+        ),
+        (Some(_), None) => return CommentType::Line,
+        (None, Some(tokens)) => (
+            find_block_comments(tokens.0, tokens.1, text, &split_lines).0,
+            find_block_comments(tokens.0, tokens.1, text, selection).0,
```

Comment:
you are essentially looking for block comments here and immedietly throwing away the result. I think there should we a second function for creating a transaction for block comment. Then you can have the following commands:
* toggle block comments (calls find_block_comment and create_block_comments_transaction)
* toggle comments: basically this function but instead of returning an enum immediately call `toggle_block_comments_as_line_fallback`, `toggle_line_comments` or `create_block_commen_transaction` (reusing the resunts of `find'block_comments` from inside this function)

Reply:
Ah that makes sense, I was a bit iffy about doing this as well and that works much better

Reply:
this is now implemented and I was also able to simplify the detection for the toggle_comments command, i then moved it into ``commands.rs`` I can move it back if you prefer though

Reply:
i think it looks great lke this. Great split: core logging in helix-core glue code in the command implementation